Adds AI Assistant/Data Modeling (#23)
dojo-modeling/dojo-docs: data-modeling.md

- `conserve` - maintain the total sum of the data before and after (e.g. for regridding population)
- `min` - take the minimum value from each bucket during regridding
- `max` - take the maximum value from each bucket during regridding
- `mean` - take the mean value from each bucket during regridding. *Note: Use interp_or_mean instead of `mean`, since it doesn't handle well when increasing the resolution of the data*
- `median` - take the median value from each bucket during regridding
- `mode`  - take the mode of each bucket during regridding. *Note: Use nearest_or_mode instead of `mode`, since it doesn't handle well when increasing the resolution of the data*
- `interp_or_mean`  - if increasing resolution, interpolate the data. If decreasing resolution, take the mean of each bucket
- `nearest_or_mode`  - if increasing resolution, take the nearest data. if decreasing resolution, take the mode of each bucket

### Threshold Node
![Threshold Node](imgs/data-modeling-threshold-node.png)

Threshold Nodes are used to select specific portions of data where some condition is true.

> Take the example of a temperature dataset where you want to filter for values above a threshold temperature. A Threshold Node that takes temperature data as input and is set to `greater_than` with the value of your threshold creates a boolean mask that is true wherever the condition was met. This can then be used to filter the original data with a [Join Node](#join-node).

Author: I changed `A filter node that takes temperature data as input...` to `A Threshold Node that takes temperature data as input...` because I think using a different name for the node here is confusing.

---
layout: default
title: Data Modeling
nav_order: 4
has_toc: true
---

# Data Modeling

> With Dojo's Data Modeling tool, users can load features from NetCDF files into a visual graphing flow, perform actions on those features, and generate new datasets from the results.

Reviewer: ```suggestion
With Dojo's Data Modeling tool, users can combine NetCDF datasets together through a simple visual programming interface to create new and insightful derived datasets.
```

## Selecting NetCDFs

When you initially land on the Data Modeling page, you'll be prompted to selected registered NetCDFs. If you have a NetCDF that you'd like to use that isn't yet registered in Dojo, follow the steps in [Data Registration](./data-registration) to register it. It will then appear in the list of NetCDF datasets.

You can select as many datasets as you'd like. Once you've selected all your datasets, click `USE THESE DATASETS` on the bottom of the page to move to the next step.

![Select NetCDFs](imgs/data-modeling-select-netcdfs.png)

## Building a graph

> On the second step, you now have to start building out your graph. Drag and drop the different nodes referenced below from the list of nodes on the right side of the screen onto the dotted surface that covers the majority of the screen. Click on the small circles at the bottom/top of each node, then move your mouse (with the line from the circle) to the circle at the bottom/top of the node you'd like to connect your selected node to. By adding nodes, described below, and connecting them together, you'll create a graph that will allow you to process data from multiple features into new dataset outputs.

Reviewer: ```suggestion
On the second step, you can start building out your data graph. Drag and drop the different nodes referenced below from the list of nodes on the right side of the screen onto the dotted surface that covers the majority of the screen. Click on the small circles at the bottom/top of each node, then move your mouse (with the line from the circle) to the circle at the bottom/top of the node you'd like to connect your selected node to. By adding nodes, described below, and connecting them together, you'll create a graph that represents the operations over the input data to create your new output datasets.
```

### Nodes:

![Node List](imgs/data-modeling-node-list.png)

#### Load Node

> ![Load Node](imgs/data-modeling-load-node.png)

Reviewer: ```suggestion
![Load Node](https://github.com/dojo-modeling/dojo-docs/assets/41205753/7726dad6-68ba-4ab6-9d2b-bf4d968c9d80)
```

> This starts with selecting Load Nodes. Each Load Node represents one feature from a dataset. You select this feature with the "Data Source" dropdown at the top of the node. You'll see features listed per dataset in this dropdown.

Reviewer: ```suggestion
This node type is how data is brought into the data modeling process. Each Load Node represents **one** feature from a dataset. You select this feature with the "Data Source" dropdown at the top of the node. You'll see features listed per dataset in this dropdown.
```

> Once you've selected a feature for a Load Node, you can choose to make it the Geo Resolution and/or the Time Resolution for your new dataset. You can only select each of these once in the entire graph, either in the Load Nodes or in the Time/Geo Resolution selectors in the sidebar.

Reviewer: ```suggestion
Once you've selected a feature, there are some additional options relating to how data regridding is handled. In a data modeling pipeline, there must be a global setting for what resolution to target when geo/temporally regridding data. You can either manually hardcode the geo/temporal resolution target (see [[#Manual Resolution]] below), you you can check one or both of the boxes on a load node to make it's resolution be the target resolution. For the entire graph, you can only have one geo resolution target, and one temporal resolution target, either in the Load Nodes or in the Time/Geo Resolution selectors in the sidebar.
```

Reviewer: don't know if I made the link for [[#Manual Resolution]] correctly

> You also need to choose a Geo and Time Aggregation function for each Load Node.

Reviewer: ```suggestion
Additionally, you must specify what type of regridding method will be used if the data in the load node needs to be regridded. These are specified with the Geo and Time Aggregation function dropdowns for each Load Node:
- `conserve` - maintain the total sum of the data before and after (e.g. for regridding population)
- `min` - take the minimum value from each bucket during regridding
- `max` - take the maximum value from each bucket during regridding
- `mean` - take the mean value from each bucket during regridding. **People should use interp_or_mean instead of this one, since it doesn't handle well when increasing the resolution of the data**
- `median` - take the median value from each bucket during regridding
- `mode`  - take the mode of each bucket during regridding. **People should use nearest_or_mode instead of this one, since it doesn't handle well when increasing the resolution of the data**
- `interp_or_mean`  - if increasing resolution, interpolate the data. If decreasing resolution, take the mean of each bucket
- `nearest_or_mode`  - if increasing resolution, take the nearest data. if decreasing resolution, take the mode of each bucket
```

<!-- Note: explain these:
'conserve', 'min', 'max', 'mean', 'median', 'mode', 'interp_or_mean', 'nearest_or_mode',
 -->

> #### Threshold Node

Reviewer: ```suggestion
#### Threshold Node
![Threshold Node](https://github.com/dojo-modeling/dojo-docs/assets/41205753/101bd10a-0a20-4954-a645-b47f60a25c4b)
Threshold nodes are typically used to select specific portions of your data where some condition is true. For example, say you have a temperature dataset, and want to filter for values above some threshold temperate; a filter node that takes the temperature data as input, and is set to `greater_than` with the value of your threshold, will create a boolean mask that is true wherever the condition was met (which could then be used to filter the original data with a [[#Join Node]]).

Threshold values can be an integer or floating point number, and the type of comparison operation (`=`, `≠`, `>`, `≥`, `<`, `≤`)  is specified with the dropdown menu.
```

> #### Join Node

Reviewer: ```suggestion
#### Join Node
![Join Node](https://github.com/dojo-modeling/dojo-docs/assets/41205753/ac855483-9e0a-44de-9620-58efde013bd3)
Join nodes are used to combine multiple nodes in the graph into a single dataset. Join nodes take as input two parent nodes, and outputs the values of those input nodes combined together. Under the hood, a join node multiplies the input datasets together. 

The most common use case for join nodes is to filter data to where some condition is true. A dataset mask can be created with a [[#Threshold Node]], and then that mask can be "joined" with the data that you want to apply the mask to. 

Less commonly, depending on the particular data operations you are trying to do, it may make sense to use a Join Node to multiple two separate datasets together to get a new dataset out the other end. E.g. if one dataset is gridded solar energy flux (watts/meter^2), and another dataset is solar panel area (meter^2) per each lat/lon point, you could multiply them together with a Join Node to get a new dataset that is the expected power production per each lat/lon grid point (assuming 100% efficiency). 

```

#### Filter By Country Node 

> With this node you can filter the data coming out of your selected features by the countries you select in the dropdown. You can select multiple countries.

Reviewer: ```suggestion
With this node you can filter the data coming out of your selected features by the countries you select in the dropdown. You can select multiple countries. The output dataset will be a copy of the input dataset, but with an extra `admin0` (i.e. country) dimension. Every country specified in the Filter By Country Node will be split into their own separate layer along the country dimension of the output. All other values not within the bounds of any of the specified countries will be set to `0`
```

> #### Reduce By Node
> <!--
> Note: explain these in context
>  'lat', 'lon', 'time', 'country'
>  -->

Reviewer: ```suggestion
#### Reduce By Node
![Reduce By Node](https://github.com/dojo-modeling/dojo-docs/assets/41205753/cd7d9451-204e-4cfe-ad18-edc60380b344)
This node can be used to aggregate a dataset across one or more dimensions. The input dataset will be `sum-reduced` along the dimensions specified by the checkboxes. For example, say you have monthly population data gridded to `0.1x0.1` lat/lon, and you would like to produce a single population curve over time. Using reduce by, and selecting `lat`, and `lon` will sum up all of the values for each time slice in the input dataset, and produce a new dataset that is just a single aggregated curve over time.

>> Note: at present Reduce By Nodes are hardcoded with the options `lat`, `lon`, `time`, and `country`, regardless of what dimensions your data has. Take care to only select dimensions that already exist on your dataset.
```

> #### Save Node
> The final step in a graph is adding a Save Node. The Save Node will provide a name and description for the output, provided by the two input fields on the node. 

Reviewer: ```suggestion
#### Save Node
![Save Node](https://github.com/dojo-modeling/dojo-docs/assets/41205753/a147372d-5eaa-4a79-811f-96861cad2f49)

The final step in a graph is adding a Save Node. The Save Node has input text fields for both name and description, which will be used to label the output dataset and save it back into dojo as a new derived dataset.
```

> ### Manual Resolution
> Below the list of nodes on the right side of the screen, you'll find two inputs where you can manually set your Geo or Time Resolution. You can use these instead of selecting one of your `Load Node` features to be Geo/Time Resolution if you want to set the resolution by hand. If you set it here, you won't be able to select any of the same type of Resolution elsewhere until you clear it. 

Reviewer: ```suggestion
### Manual Resolution
![Manual Resolution](https://github.com/dojo-modeling/dojo-docs/assets/41205753/63787be6-66e8-435d-9cfc-cfcfe5226616)

Below the list of nodes on the right side of the screen, you'll find two inputs where you can manually set your Geo or Time Resolution. You can use these instead of selecting one of your `Load Node` features to be Geo/Time Resolution if you want to set the resolution by hand. If you set it here, you won't be able to select any of the same type of Resolution elsewhere until you clear it. 

>> Note that somewhere in the data modeling graph, you must set a resolution for both geo and temporal regridding. That could be either here in the side bar, on individual Load Nodes with the checkbox, or one of each
```
